libraries/AC_Fence/AC_Fence.cpp

```diff
@@ -155,227 +159,324 @@ AC_Fence::AC_Fence()
 #endif
     _singleton = this;
     AP_Param::setup_object_defaults(this, var_info);
+    if (_enabled) {
+        _enabled_fences = _configured_fences.get() & ~AC_FENCE_TYPE_ALT_MIN;
+    }
+}
+
+// get a user-friendly list of fences
+void AC_Fence::get_fence_names(uint8_t fences, ExpandingString& msg)
+{
+    if (!fences) {
+        return;
+    }
+    static const char* FENCE_NAMES[] = {
+        "Max Alt",
+        "Circle",
+        "Polygon",
+        "Min Alt",
+    };
+    uint8_t i = 0;
+    uint8_t nfences = 0;
+    while (fences !=0) {
+        if (fences & 0x1) {
+            if (nfences > 0) {
+                if (!(fences & ~1U)) {
+                    msg.printf(" and ");
+                } else {
+                    msg.printf(", ");
+                }
+            }
+            msg.printf("%s", FENCE_NAMES[i]);
+            nfences++;
+        }
+        fences >>= 1;
+        i++;
+    }
+    msg.printf(" fence");
+    if (nfences>1) {
+        msg.printf("s");
+    }
+}
+
+// print a message about the passed in fences
+void AC_Fence::print_fence_message(const char* message, uint8_t fences) const
+{
+    if (!fences) {
+        return;
+    }
+
+    char msg[MAVLINK_MSG_STATUSTEXT_FIELD_TEXT_LEN+1];
+    ExpandingString e(msg, MAVLINK_MSG_STATUSTEXT_FIELD_TEXT_LEN+1);
+    AC_Fence::get_fence_names(fences, e);
+    GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "%s %s", e.get_writeable_string(), message);
+}
+
+// should be called @10Hz to handle loading from eeprom
+void AC_Fence::update()
+{
+    _poly_loader.update();
+    // if someone changes the parameter we want to enable or disable everything
+    if (_enabled != _last_enabled || _auto_enabled != _last_auto_enabled) {
+        // reset the auto mask since we just reconfigured all of fencing
+        _auto_enable_mask = AC_FENCE_ALL_FENCES;
+        _last_enabled = _enabled;
+        _last_auto_enabled = _auto_enabled;
+        if (_enabled) {
+            _enabled_fences = _configured_fences.get() & ~AC_FENCE_TYPE_ALT_MIN;
+        } else {
+            _enabled_fences = 0;
+        }
+    }
+#ifdef AC_FENCE_DEBUG
+    static uint32_t last_msg_count = 0;
+    if (get_enabled_fences() && last_msg_count++ % 10 == 0) {
+        print_fence_message("active", get_enabled_fences());
+        print_fence_message("breached", get_breaches());
+    }
+#endif
 }
 
 /// enable the Fence code generally; a master switch for all fences
-void AC_Fence::enable(bool value)
+uint8_t AC_Fence::enable(bool value, uint8_t fence_types, bool update_auto_mask)
 {
+    uint8_t fences = _configured_fences.get() & fence_types;
+    uint8_t enabled_fences = _enabled_fences;
+    if (value) {
+        enabled_fences |= fences;
+    } else {
+        enabled_fences &= ~fences;
+    }
+
+    // fences that were manually changed are no longer eligible for auto-enablement or disablement
```

Comment:
yeah that's intentional - if the user manually alters a fence's state via mavlink we honour that to avoid surprises. I think there are even some tests that rely on this behaviour